### PR TITLE
Avoid redundant looping in property metadata

### LINKED ISSUE #29642: Performance regression in model building in Entity Framework Core 7
Hi,

I have just migrated an application to Entity Framework Core 7 and noticed that executing the migrations for a new database takes twice or even three times longer than with Entity Framework Core 6. As far as I could see, adding interceptors, the actual SQL statements perform roughly the same, it's the time between migrations that takes longer (I guess building the operations, but I didn't go that deep).

I created a repo with a very simple project to showcase this scenario: https://github.com/MaQy/EfCoreMigrationsBenchmark

Even with just 5 migrations, 4 of them pretty small, the issue can be seen according to the results I got:

| Method |  Runtime |       Mean |    Error |   StdDev |
|------- |--------- |-----------:|---------:|---------:|
|   Test | .NET 6.0 |   547.9 ms | 302.9 ms | 200.4 ms |
|   Test | .NET 7.0 | 1,412.9 ms | 421.2 ms | 278.6 ms |

### Include provider and version information

EF Core version: 7.0
Database provider: Microsoft.EntityFrameworkCore.SqlServer
Target framework: .NET 7.0
Operating system: Windows 11 21H2
IDE: Visual Studio 2022 17.4

### PR DESCRIPTION
**Pull request by AndriySvyryd**
Fixes #29642

### CODE REVIEW

File: src/EFCore.Design/Scaffolding/Internal/CSharpRuntimeModelCodeGenerator.cs [hunk 1]
```diff
@@ -858,29 +858,37 @@ private void Create(IEntityType entityType, CSharpRuntimeAnnotationCodeGenerator
         var principalProperty = property;
         for (var i = 0; i < 10000; i++)
```
Comment (ajcvickers) [inline src/EFCore.Design/Scaffolding/Internal/CSharpRuntimeModelCodeGenerator.cs:859]: Is 10,000 really the appropriate number here?

**Comment by ajcvickers** [comment]
Can you say a bit more about what is happening here? What are the situations where the 10,000 limit will be hit? Also, which tests cover these cases?

**Comment by AndriySvyryd** [comment]
The code is traversing FKs. The 10000 limit will be hit if there's an FK cycle in the model. These models are invalid and will fail validation, but the current code runs before validation, so it shouldn't fail before validation runs.

Tests:
https://github.com/dotnet/efcore/blob/f7328757bb8120ed3284a6a136a40bebe5e891a3/test/EFCore.Tests/Infrastructure/ModelValidatorTest.cs#L475
https://github.com/dotnet/efcore/blob/f7328757bb8120ed3284a6a136a40bebe5e891a3/test/EFCore.Relational.Tests/Infrastructure/RelationalModelValidatorTest.cs#L578

File: src/EFCore.Design/Scaffolding/Internal/CSharpRuntimeModelCodeGenerator.cs (hunk 1 above)
Comment (AndriySvyryd) [reply src/EFCore.Design/Scaffolding/Internal/CSharpRuntimeModelCodeGenerator.cs:859]: It should be the length of the longest FK chain allowable in the model

**Comment by ajcvickers** [comment]
The customer report didn't fail in validation. I get this was likely a bug, but it makes me nervous that we still have a very high limit such that some other bug could cause very slow model building without failing in validation. Should we not just fail early when we exhaust the loop, rather than assuming model building will fail?

**Comment by AndriySvyryd** [comment]
> Should we not just fail early when we exhaust the loop, rather than assuming model validation will fail?

Sure, but only in main

**Comment by ajcvickers** [comment]
Only in main seems reasonable.

**Review by ajcvickers — APPROVED** [review]

**Pull request merged**